### PR TITLE
Account Creation With Getcode

### PR DESCRIPTION
**Pull request by PrettyCoolPattern**
Latest EdenMoogles with 100% working verify, recapcha, & getcode form query + response.

Query is to dspdb database for table code at column code.

Line 75 of api/vi/accounts.jsx to modify:

`    const servercodequery = 'SELECT code FROM code;';`

### CODE REVIEW

**Comment by PrettyCoolPattern** [comment]
I was considering asking about adding the test server to the header refer check and !verify process. 

Would this be of benefit to testing?

Another note to consider is the recapcha may need adjusting for the test.edenxi testing as wel.

**Comment by PrettyCoolPattern** [comment]
Also, when we settle on the db query location we can update the readme, and sql setup files to include this provision.

File: api/v1/accounts.js [hunk 1]
```diff
@@ -22,7 +22,6 @@ const validate = (req, res, next) => {
     }
   });
 };
-
```
Comment (InoUno) [inline api/v1/accounts.js:25]: Please re-add this newline for readability.

File: api/v1/accounts.js [hunk 2]
```diff
@@ -72,6 +72,10 @@ router.post('/register', (req, res) => {
    * http://play.edenxi.com/tools
    */
   try {
+    const servercodequery = 'SELECT code FROM code;';
+    const servercode = await req.app.locals.query(servercodequery);
+    const codebool = [servercode === code];
```
Comment (InoUno) [inline api/v1/accounts.js:77]: Why is this a boolean array?

File: api/v1/accounts.js [hunk 3]
```diff
@@ -83,8 +87,8 @@ router.post('/register', (req, res) => {
       password.length >= 6 && password.length <= 15,
       !!username,
       !!verify,
+      servercode === code,
     ];
-
```
Comment (InoUno) [inline api/v1/accounts.js:87]: Keep newline for readability

File: api/v1/accounts.js [hunk 4]
```diff
@@ -102,7 +117,7 @@ router.post('/register', (req, res) => {
           body: {},
           json: true,
         },
-        async (error, resp) => {
```
Comment (InoUno) [inline api/v1/accounts.js:105]: Why was this removed? The `resp` argument should not be removed from the parameters list, since it's used within the function body.

File: api/v1/index.js [hunk 5]
```diff
@@ -2,12 +2,12 @@ const { Router } = require('express');
 
 const router = Router();
 
-// const accounts = require('./accounts');
+const accounts = require('./accounts');
 const chars = require('./chars');
 const items = require('./items');
 const misc = require('./misc');
 
-// router.use('/accounts', accounts);
```
Comment (InoUno) [inline api/v1/index.js:10]: This still needs to be commented out, since the account management section should remain disabled for now.

File: client/src/components/accounts/login.jsx [hunk 6]
```diff
@@ -34,11 +34,15 @@ export default ({ error, login, changePage }) => {
         <Button
           floated="right"
           primary
+          disabled
           onClick={() => login(username, password)}
         >
-          Login
+          Login (coming soon)
```
Comment (InoUno) [inline client/src/components/accounts/login.jsx:40]: Don't think we want the `(coming soon)` part added.

File: client/src/components/accounts/login.jsx [hunk 7]
```diff
@@ -34,11 +34,15 @@ export default ({ error, login, changePage }) => {
...
         </Button>
-        <Button size="mini" basic onClick={changePage}>
+        <Button size="mini" disabled basic onClick={changePage}>
+          Account recovery(coming soon)
+        </Button>{' '}
```
Comment (InoUno) [inline client/src/components/accounts/login.jsx:44]: Remove excess space added (the `{' '}` after button close tag).

File: client/src/components/accounts/login.jsx [hunk 8]
```diff
@@ -34,11 +34,15 @@ export default ({ error, login, changePage }) => {
...
           primary
+          disabled
           onClick={() => login(username, password)}
         >
-          Login
+          Login (coming soon)
         </Button>
-        <Button size="mini" basic onClick={changePage}>
+        <Button size="mini" disabled basic onClick={changePage}>
+          Account recovery(coming soon)
```
Comment (InoUno) [inline client/src/components/accounts/login.jsx:43]: Same comment about coming soon.

File: client/src/components/accounts/register.jsx [hunk 9]
```diff
@@ -46,17 +48,19 @@ export default ({ error, register, changePage, verify, verified }) => {
...
-          primary
           disabled={!verified}
           onClick={() => register(inputGroup)}
         >
           Register
         </Button>
-        <Button size="mini" basic onClick={changePage}>
-          Existing Account
+        <Button size="mini" disabled basic onClick={changePage}>
+          Existing Account (coming soon)
```
Comment (InoUno) [inline client/src/components/accounts/register.jsx:63]: Same thing about coming soon.

File: client/src/components/tools.jsx [hunk 10]
```diff
@@ -17,6 +17,7 @@ const Tools = () => {
   const item = params.get('item');
   const stack = params.get('stack');
   const player = params.get('player');
+  const yells = params.get('yells');
```
Comment (InoUno) [inline client/src/components/tools.jsx:20]: There is no query param called `yells`. This will never be set.

File: client/src/components/tools.jsx [hunk 11]
```diff
@@ -32,36 +33,36 @@ const Tools = () => {
...
       <div className="gm_tools-content">
         <Menu pointing>
-          <Menu.Item
-            disabled
-            active={selected === 'account'}
-            onClick={updateTab('account')}
-          >
-            User Management
-          </Menu.Item>
+          {' '}
```
Comment (InoUno) [inline client/src/components/tools.jsx:36]: Remove excess space added.

File: client/src/components/tools.jsx [hunk 12]
```diff
@@ -32,36 +33,36 @@ const Tools = () => {
...
             onClick={updateTab('online')}
           >
             Who's Online
           </Menu.Item>
+          <Menu.Item
+            active={selected === 'account'}
+            onClick={updateTab('account')}
+          >
+            User Management
+          </Menu.Item>
```
Comment (InoUno) [inline client/src/components/tools.jsx:48]: Why was this moved to after `Who's Online`?
If we want to reorder it, it should be done in a separate PR.

File: client/src/components/tools.jsx [hunk 13]
```diff
@@ -32,36 +33,36 @@ const Tools = () => {
...
           <Menu.Item active={selected === 'items'} onClick={updateTab('items')}>
             Item Search
           </Menu.Item>
           <Menu.Item active={selected === 'chars'} onClick={updateTab('chars')}>
             Player Search
-          </Menu.Item>
+          </Menu.Item>{' '}
```
Comment (InoUno) [inline client/src/components/tools.jsx:54]: Remove excess space.

File: client/src/components/tools.jsx [hunk 14]
```diff
@@ -32,36 +33,36 @@ const Tools = () => {
...
           <Menu.Item active={selected === 'yells'} onClick={updateTab('yells')}>
             Yells
           </Menu.Item>
         </Menu>
-        {/* {selected === 'account' && <Accounts />} */}
         {selected === 'online' && <OnlineList />}
+        {selected === 'account' && <Accounts />}
```
Comment (InoUno) [inline client/src/components/tools.jsx:60]: Keep it commented out for now, as we don't want users to use it yet.

File: client/src/components/tools.jsx [hunk 15]
```diff
@@ -32,36 +33,36 @@ const Tools = () => {
...
         {(item || selected === 'items') && (
           <Itemsearch itemname={item} itemstack={stack} />
         )}
         {(player || selected === 'chars') && <Playersearch charname={player} />}
-        {selected === 'yells' && <YellTab />}
+        {yells || (selected === 'yells' && <YellTab />)}
```
Comment (InoUno) [inline client/src/components/tools.jsx:65]: Remove the `yells ||` check again, since that query param is never set.

File: client/src/components/accounts/index.jsx [hunk 16]
```diff
@@ -138,7 +138,12 @@ class Account extends React.Component {
     } else {
       delete errors.confirmEmail;
     }
-
+    if (info.code == null) {
+      errors.code =
+        'You need a registration code. Obtain one from an established player, or try asking on our Discord.';
+    } else {
+      delete errors.code;
+    }
```
Comment (InoUno) [inline client/src/components/accounts/index.jsx:146]: Add newline after this for readability (like it had after the email check).

File: api/v1/accounts.js (hunk 1 above)
Comment (PrettyCoolPattern) [reply api/v1/accounts.js:25]: Omg I didn't mean to resolve this, just about to get started on working on these, thank you for the overlooks.

File: api/v1/index.js (hunk 5 above)
Comment (PrettyCoolPattern) [reply api/v1/index.js:10]: Accounts will be back in with this PR, login and account creation are handled on the same page of /accounts

File: api/v1/accounts.js (hunk 4 above)
Comment (PrettyCoolPattern) [reply api/v1/accounts.js:105]: Removed because I brought the request and response to the router post section of the script to allow for inquiry of the code prior to the inquiry for username taken.
line 54:
router.post('/register', async (req, res) => {

File: api/v1/accounts.js (hunk 2 above)
Comment (PrettyCoolPattern) [reply api/v1/accounts.js:77]: I wasn't certain if the structure was causing conflicts. It ended up like this but works as three constants for its three separate functions.

File: client/src/components/tools.jsx (hunk 14 above)
Comment (PrettyCoolPattern) [reply client/src/components/tools.jsx:60]: As with my prior comment, Accounts will be where people register as it is coded as a shared page with login.

File: client/src/components/tools.jsx (hunk 15 above)
Comment (PrettyCoolPattern) [reply client/src/components/tools.jsx:65]: I added the query param actually and returned the tools page to priority to Accounts as the opening tab.

File: client/src/components/accounts/index.jsx (hunk 16 above)
Comment (PrettyCoolPattern) [reply client/src/components/accounts/index.jsx:146]: ✔

File: api/v1/accounts.js (hunk 1 above)
Comment (PrettyCoolPattern) [reply api/v1/accounts.js:25]: ✔

File: client/src/components/tools.jsx (hunk 13 above)
Comment (PrettyCoolPattern) [reply client/src/components/tools.jsx:54]: ✔

File: api/v1/accounts.js (hunk 3 above)
Comment (PrettyCoolPattern) [reply api/v1/accounts.js:87]: ✔

File: client/src/components/accounts/login.jsx (hunk 7 above)
Comment (PrettyCoolPattern) [reply client/src/components/accounts/login.jsx:44]: ✔

File: client/src/components/tools.jsx (hunk 12 above)
Comment (PrettyCoolPattern) [reply client/src/components/tools.jsx:48]: Good eye on that one mate.

**Comment by bkesecker** [comment]
Please remove the !getcode feature.  The more accessible it is, the more it can be abused.  I should also mention, anything on the website that touches the database needs to be throttled.
